### PR TITLE
adding bounds check for advanced indexing

### PR DESCRIPTION
**Pull request by ipdemes**
addressing issue #31

### CODE REVIEW

**Comment by magnatelee** [comment]
@ipdemes is this a complete PR? I don't see any changes on the Python side to make the exceptions re-raised.

File: src/cunumeric/index/zip.h [hunk 1]
```diff
@@ -45,6 +45,9 @@ class ZipTask : public CuNumericTask<ZipTask> {
 
 constexpr coord_t compute_idx(coord_t index, coord_t dim)
 {
+  // bounds checking
+  if ((index < 0 && index + dim < 0) || (index >= 0 && index >= dim))
+    throw legate::TaskException("index is out ouf bounds in index array");
```
Comment (manopapad) [inline src/cunumeric/index/zip.h:50]: Typo: "out of bounds"

**Comment by ipdemes** [comment]
> @ipdemes is this a complete PR? I don't see any changes on the Python side to make the exceptions re-raised.

That is right, I didn't push changes to the python code

**Comment by manopapad** [comment]
I think we should throw the standard Python `IndexError` exception here, as NumPy does. We don't need a custom `CuNumericError` in this case.

```
>>> import numpy as np
>>> x = np.ones((4,))
>>> x[[0,1,2,3,4]]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: index 4 is out of bounds for axis 0 with size 4
```

**Comment by ipdemes** [comment]
@manopapad makes sense. Will change this

File: cunumeric/exception.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+class IndexError(Exception):
```
Comment (manopapad) [inline cunumeric/exception.py:17]: I think `IndexError` is a builtin python class, you can just reuse that.

Comment (ipdemes) [reply cunumeric/exception.py]: That is right

**Review by manopapad — APPROVED** [review]

File: src/cunumeric/index/zip.h [hunk 3]
```diff
@@ -45,6 +45,9 @@ class ZipTask : public CuNumericTask<ZipTask> {
 
 constexpr coord_t compute_idx(coord_t index, coord_t dim)
```
Comment (magnatelee) [inline src/cunumeric/index/zip.h:46]: the variable name `dim` is very confusing. I believe you meant `extent`, right? `dim` seems synonymous with `axis` in other contexts, so let's change this.

File: src/cunumeric/index/zip.h [hunk 4]
```diff
@@ -45,6 +45,9 @@ class ZipTask : public CuNumericTask<ZipTask> {
...
 {
+  // bounds checking
+  if ((index < 0 && index + dim < 0) || (index >= 0 && index >= dim))
```
Comment (magnatelee) [inline src/cunumeric/index/zip.h:49]: I'd actually rewrite this function to something like this:
```
coord_t sanitized_index = index < 0 ? index + dim : index;
if (sanitized_index < 0 || sanitized_index >= extent) throw ...
return sanitized_index;
```

**Comment by magnatelee** [comment]
@ipdemes merge this with the latest branch-22.07, then you won't see the failure on 2 GPUs anymore.

**Comment by ipdemes** [comment]
> @ipdemes merge this with the latest branch-22.07, then you won't see the failure on 2 GPUs anymore.

done

**Review by magnatelee — APPROVED** [review]

**Pull request merged**